### ISSUE TITLE
Compiler allow wrong implicit convert from zero (0) to Enum in method argument without emit an error or warning!

### ISSUE THREAD

**Issue by MauNguyenVan** [dotnet/roslyn#68127]
**[VS 2022 17.5.5] [.NET 7.0.203]**: 

**Reproduce code**:
`Program.cs` file.
```csharp

GetEnum(0);// Bad, cause VS allowed, do not emit an error "CS1503: Cant not convert from int to CustomEnum"
GetEnum(0u);// Bad, cause VS allowed, do not emit an error "CS1503: Cant not convert from uint to CustomEnum"
GetEnum(0b0);// Bad, cause VS allowed, do not emit an error "CS1503: Cant not convert from int to CustomEnum"
GetEnum(0x0);// Bad, cause VS allowed, do not emit an error "CS1503: Cant not convert from int to CustomEnum"
GetEnum(0f);// Bad, cause VS allowed, do not emit an error "CS1503: Cant not convert from float to CustomEnum"
GetEnum(0d);// Bad, cause VS allowed, do not emit an error "CS1503: Cant not convert from double to CustomEnum"
GetEnum(0m);// Bad, cause VS allowed, do not emit an error "CS1503: Cant not convert from decimal to CustomEnum"
GetEnum(0L);// Bad, cause VS allowed, do not emit an error "CS1503: Cant not convert from long to CustomEnum"
GetEnum(0ul);// Bad, cause VS allowed, do not emit an error "CS1503: Cant not convert from ulong to CustomEnum"

GetEnum((byte)0);// Bad, cause VS allowed, do not emit an error "CS1503: Cant not convert from byte to CustomEnum"
GetEnum((sbyte)0);// Bad, cause VS allowed, do not emit an error "CS1503: Cant not convert from byte to CustomEnum"
GetEnum((int)0);// Bad, cause VS allowed,  do not emit an error "CS1503: Cant not convert from int to CustomEnum"
GetEnum((int)0);// Bad, cause VS allowed, do not emit an error "CS1503: Cant not convert from int to CustomEnum"
GetEnum((float)0);// Bad, cause VS allowed, do not emit an error "CS1503: Cant not convert from int to CustomEnum"

GetEnum(1);// Ok 
GetEnum(0b01);// Ok 
GetEnum((int)1);// Ok 

static void GetEnumError(CustomEnum customEnum =0)// Bad , do not emit an Error: Error	CS1750	A value of type 'int' cannot be used as a default parameter because there are no standard conversions to type 'CustomEnum'
{
    Console.WriteLine("Enum error is " + customEnum);
}

static void GetEnumErrorOK(CustomEnum customEnum = 1)// OK , emited an Error: Error	CS1750	A value of type 'int' cannot be used as a default parameter because there are no standard conversions to type 'CustomEnum'
{
    Console.WriteLine("Enum error is " + customEnum);
}

static void GetEnum(CustomEnum customEnum)
{
    Console.WriteLine("Enum error is " + customEnum);
}
enum CustomEnum
{
    A = 1,
    B = 2,
    C = 3
}

```
 

 [see my code at sharplab](https://sharplab.io/#v2:C4LgTgrgdgNAJiA1AHwAICYCMBYAUKgZgAIMiBhIgbzyNpPXSrxrtYHEBTYAUSggFsAFAAYAlAG4A9JKIAhAIZwYRAMbyIAZw5EAagGUi8gDZGA9gHcOSonFNEop4EQ78Alk/lRnYMKbBEAIjI9TABWYQIQck8nBycVUygANw4wJwAzX34iVygnYDsyTQL+XgEAllZaTh4+IWEICWk5RWU1TW19QxMLK2Vbe0dnNw8vVN9/IJDwyOi8wfjElLSiTNNsiFz8wuL1sv4K3Cq6Gv2RACMxKRkFa3atXQNjM0trAbjh90Mxnz9A4LCESiZBiC1US1SGSyOXmBXIu1KdUOx2qXDOwgAHldmrc2uoHl1nr03nYPi4vp5vBN/tMgXNYkMEslIatoVsiHCihoSvtkSjTnUROkmjdWqp8Z0nj1Xv1SUNyaMqX8poDZiD5h8mcsoetVmZ5Nt4dy9kjKlUBQIRHARS07hLHt0Xn0bHKnArvkrJgCZsDQZqISs1tlbBBzkZtJyEbyzew0YLhPwbbjxR0HUSZS6we7KeNld66eqGYtmYHoXAOCpXPxjBydsbEeUYyc45bhAAZJNi+6Sx3E2VZkYe3Ne2lqv2MgM67JmKAAc1rRp5pqO/Jb9QgRk7dtThOlzve8sHOd+I9Vvo1E5LU6IG8S88j9ejuCbqNqlsE5wAnsAOKJsaLtwJKUnRJAcKR+akVR9ekwS1FkgyIL8fwXLkl0bFdjgtIRBA0JDf3/W08R3YC+0zMkjwgvNR3PItwSvVldTwlCo2XFEiCwwRBC2P8tyIoDewzMjD3Az0aTPGD/XohD2QfNCDhfdi1047iCOTbs0z3UDyJE4cxOgwtYMnBjshkus5L5TClMEdJ9WAHjrkIlN+PTfdXU+RVdKggtx2LbVjJhQ1UJNdCFOkDjMBtAB5ABrIhQskDjhEuCKHJiuKMKqMKrO4lLmjSs0AF8zVQTAADYSAAFkUt9ShPQQgobbIVBYgQiAAXjEHFFCIfttMVbgTyiAaJgASABAB2cIRoAQSIJJjAgbRTHSDlPwAB20AByLZNvFKAPnObQOjgQwNEMGwOHSdQjCcNb5DAeR+C4VJEIre1gAAC1SbR7u0BwiG5Tw4Huk64LADRXESM64WAdatoa/ZNrNagMtYEqAE5BACfZRNcM6AiIRBVBaxNxEK5gMukErytQKqsOGvxIoAaXqknicfOp2qICK8qZnr3KsD0GbAIaTzGkJJuEGa5oWpaVthjaiG2vJdrUfahkOm8tBO+QzvkC6ro3W77se57/EO9TPu+wwwD+uxAagYGwFBgMIahhcFfhkmkcpyQUbYqnMEx7HOd0vHAkJ9m5IkM1pCK58MupyrqrOBHOeajmBFEZGFIxrGcbD/HI4z6OyYy+PWA4Tm04EHPUboWa2u5mAFNkLn0Bb+vaAoJuCHJ3ACqAA==) 

**Expected Behavior**:

```csharp
GetEnum(0);//  Will emit an error `Error	CS1503	Argument 1: cannot convert from 'int' to 'CustomEnum'`

static void GetEnumError(CustomEnum customEnum = 0)// Will emit  an Error: Error	CS1750	A value of type 'int' cannot be used as a default parameter because there are no standard conversions to type 'CustomEnum'
{
    Console.WriteLine("Enum error is " + customEnum);
}
```

**Actual Behavior**:
```csharp
GetEnum(0); // Bad, It allow without no error or warning

static void GetEnumError(CustomEnum customEnum =0)// Bad , do not emit an Error: Error	CS1750	A value of type 'int' cannot be used as a default parameter because there are no standard conversions to type 'CustomEnum'
{
    Console.WriteLine("Enum error is " + customEnum);
}
```

**Comment by svick** [comment]
Most of these correct according to [the spec](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/conversions#1024-implicit-enumeration-conversions):

> An implicit enumeration conversion permits a *constant_expression* with any integer type and the value zero to be converted to any *enum_type* [...].

Some of them (at least the floating point ones) are an intentional violation of the spec by the compiler, to maintain compatibility:

https://github.com/dotnet/roslyn/blob/41532383589ea5dccc7a3996656628ed9ac92e6a/src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs#L1371-L1375

(Note that the comment is outdated, I think the spec was loosened since it was written.)

So I don't think there's actually anything to do here, unless there is a particular reason to reevaluate the decision to violate the spec.

**Comment by jcouv** [comment]
Thanks @svick for tracking this down. Seeing that the spec allows for this in a number of cases and the remaining cases are supported for historical reasons, I'll close this as by-design. 
Changing this (even with a warning) would break a lot of code, so it would need a very strong justification to even contemplate.
